bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java

```diff
@@ -0,0 +1,506 @@
+            return SslContext.newClientContext();
+          } catch (SSLException e) {
+            throw new IllegalStateException("Could not create an ssl context.", e);
+          }
+        }
+      };
+
+  private final static Logger LOG = new Logger(BigtableSession.class);
+
+  public static final String GRPC_EVENTLOOP_GROUP_NAME = "bigtable-grpc-elg";
+
+  /** Number of milliseconds to wait for a termination before trying again. */
+  public static final long CHANNEL_TERMINATE_WAIT_MS = 5000;
+
+  protected static EventLoopGroup createDefaultEventLoopGroup() {
+    return new NioEventLoopGroup(0, new ThreadFactoryBuilder().setDaemon(true)
+      .setNameFormat(GRPC_EVENTLOOP_GROUP_NAME + "-%d").build());
+  }
+
+  protected static ScheduledExecutorService createDefaultRetryExecutor() {
+    // Creating the channel options is slow.  Specifically, the credential creation is slow.
+    // There are some slow operations in configureBigtableOptions as well, specifically
+    // the InetAddress.getByName.  The slowest operation by far is the credential creation,
+    // so do it in parallel.
+    //
+    // The retryExecutor is available for use.  While it's a bit weird to use retryExecutor, but
+    // since it's already built and not used elsewhere, might as well use it here.
+    return Executors.newScheduledThreadPool(
+        RETRY_THREAD_COUNT,
+        new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat(RETRY_THREADPOOL_NAME + "-%d")
+            .build());
+  }
+
+  private BigtableClient client;
+  private BigtableTableAdminClient tableAdminClient;
+  private BigtableClusterAdminClient clusterAdminClient;
+
+  private final BigtableOptions options;
+  private final ExecutorService batchPool;
+  private final EventLoopGroup elg;
+  private final ScheduledExecutorService scheduledRetries;
+
+  public BigtableSession(BigtableOptions options, ExecutorService batchPool) {
+    this(options, batchPool, createDefaultEventLoopGroup(), createDefaultRetryExecutor());
+  }
+
+  public BigtableSession(BigtableOptions options, ExecutorService batchPool, EventLoopGroup elg,
+      ScheduledExecutorService scheduledRetries) {
+    this.elg = elg;
+    this.batchPool = batchPool;
+    this.options = options;
+    LOG.info("Opening connection for project %s on data host %s, "
+        + "table admin host %s, using transport %s.", options.getProjectId(),
+      options.getDataHost(), options.getTableAdminHost(),
+      TransportOptions.BigtableTransports.HTTP2_NETTY_TLS);
+    this.scheduledRetries = scheduledRetries;
+  }
+
+  protected TransportOptions createTransportOptions(InetAddress host) {
+    return new TransportOptions(
+        TransportOptions.BigtableTransports.HTTP2_NETTY_TLS,
+        host,
+        options.getPort(),
+        SSL_CONTEXT_FACTORY,
+        elg);
+  }
+
+  public synchronized BigtableClient getDataClient() {
+    if (this.client == null) {
```

Comment:
Sounds good. Note that this might be moving blocking time from when the session is created to the first operation, depending when getDataClient is first called.